app/multitenant/dynamo_collector.go

```diff
@@ -419,9 +434,63 @@ func (c *dynamoDBCollector) Add(ctx context.Context, rep report.Report) error {
 		dynamoConsumedCapacity.WithLabelValues("PutItem").
 			Add(float64(*resp.ConsumedCapacity.CapacityUnits))
 	}
-	return err
+	if err != nil {
+		return err
+	}
+
+	if rep.Shortcut {
+		err := c.nats.Publish(userid, []byte(s3Key))
+		natsRequests.WithLabelValues("Publish", errorCode(err)).Add(1)
+		if err != nil {
+			log.Errorf("Error sending shortcut report: %v", err)
+		}
+	}
+
+	return nil
 }
 
-func (c *dynamoDBCollector) WaitOn(context.Context, chan struct{}) {}
+func (c *dynamoDBCollector) WaitOn(ctx context.Context, waiter chan struct{}) {
+	userid, err := c.userIDer(ctx)
+	if err != nil {
+		log.Errorf("Error getting user id in WaitOn: %v", err)
+		return
+	}
+
+	sub, err := c.nats.SubscribeSync(userid)
+	natsRequests.WithLabelValues("SubscribeSync", errorCode(err)).Add(1)
+	if err != nil {
+		log.Errorf("Error subscribing for shortcuts: %v", err)
+		return
+	}
+
+	c.waitersLock.Lock()
+	c.waiters[watchKey{userid, waiter}] = sub
+	c.waitersLock.Unlock()
 
-func (c *dynamoDBCollector) UnWait(context.Context, chan struct{}) {}
+	go func() {
+		_, err := sub.NextMsg(natsTimeout)
+		natsRequests.WithLabelValues("NextMsg", errorCode(err)).Add(1)
+		log.Debugf("NextMsg error: %v", err)
+		close(waiter)
+	}()
+}
+
+func (c *dynamoDBCollector) UnWait(ctx context.Context, waiter chan struct{}) {
+	userid, err := c.userIDer(ctx)
+	if err != nil {
+		log.Errorf("Error getting user id in WaitOn: %v", err)
+		return
+	}
+
+	c.waitersLock.Lock()
+	key := watchKey{userid, waiter}
+	sub := c.waiters[key]
+	delete(c.waiters, key)
+	c.waitersLock.Unlock()
```

Comment:
I feel like there should already be a structure for a lock-protected dictionary. But I guess not having generics means that it's easier to just write it ourselves?